Resdnn inference

frheault opened the pull request:
This PR is replacing #2123 and contains only the inference, no training.

It contains a model for [1] to predict fODF from raw DWI signal (as SH). It is not a local model like CSD or DTI, but the class has a fit function to make it easy to use on whole volumes.

This PR contains a fetcher for the weights, the predict/fit functions in the class, an example and simple tests.

[1] Nath, V., Schilling, K. G., Parvathaneni, P., Hansen, C. B., Hainline, A. E., Huo, Y., ... & Stepniewska, I. (2019). Deep learning reveals untapped information for local white-matter fiber reconstruction in diffusion-weighted MRI. Magnetic resonance imaging, 62, 220-227.

PS: I tried to run pytest, but I had difficulties. How can I efficiently check if the docstring and tests are right and up to the standard?

skoudoro commented:
Wow, Thank you for doing this!  I started this morning to reshape, rebase, etc #2123.  Should I continue? Otherwise, I can do a PR on your PR for some changes. 

Concerning this PR, you need to put Tensorflow as an optional dependency and add some tests.

skoudoro commented:
> I tried to run pytest, but I had difficulties. How can I efficiently check if the docstring and tests are right and up to the standard?

Can you tell us more ?

pep8speaks commented:
Hello @frheault, Thank you for updating !





* In the file [`dipy/nn/histo_resdnn.py`](https://github.com/dipy/dipy/blob/0bfcdfa02aea65e0642d390b2839d7fa5257e332/dipy/nn/histo_resdnn.py):

> [Line 139:81](https://github.com/dipy/dipy/blob/0bfcdfa02aea65e0642d390b2839d7fa5257e332/dipy/nn/histo_resdnn.py#L139): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 80 characters)

frheault commented:
@skoudoro I think this PR should be the main one, I removed the training and changed the organization of the files.

As for the test, I can't seem to run pytest on my computer.
I can run `python dipy/nn/test/test_histo_resdnn.py` but `pytest dipy/nn/test/test_histo_resdnn.py` crashes immediately and I get a numpy error about compatibility... (Weird since I can easily run the test code manually)
![image](https://user-images.githubusercontent.com/10820351/154588738-4a6ce20d-29be-4248-9fdb-8c4be6a9e64d.png)

arokem commented:
@frheault : this is often the sign of a borked numpy installation. What version of numpy are you running there? What happens if you upgrade numpy in that env?

skoudoro commented:
Also, have you seen my PR in your branch ? https://github.com/frheault/dipy/pull/1 

Just doing some small correction

Otherwise, I agree with @arokem, Can you share your file to see if we reproduce the same issue.

skoudoro commented:
I do not succeed to generate your issue. When I do `pytest -svv dipy/nn/tests/test_histo_resdnn.py`, all tests succeed, no crash.

![Capture d’écran 2022-02-18 à 17 14 07](https://user-images.githubusercontent.com/23106443/154720818-533e8a62-5afb-417e-98df-346890b2a116.png)

Also, do not forget to look and merge at some of my corrections in your PR https://github.com/frheault/dipy/pull/1

frheault commented:
@skoudoro I will merge your PR, but keep the nibabel load/save.
I personally dislike quite a lot the load_nifti and save_nifti functions.
Thank you for all others fixes!

skoudoro commented:
> I personally dislike quite a lot the load_nifti and save_nifti functions.

Also, the goal is to keep all tutorials consistent, using the same functionality for now.

skoudoro commented:
> @skoudoro : what version of numpy are you working with?

just created a new env. So `tensorflow==2.8.0`, `numpy==1.22.2`

skoudoro commented on dipy/nn/histo_resdnn.py:
```diff
@@ -0,0 +1,245 @@
+
+        return self.model.predict(x_test)
+
+    def fit(self, data, gtab, mask=None):
```
This needs some thinking. I understand what you are trying to do but there is a clear conflict with `fit` function in the model of this module (`dipy.nn.model`). Look [here](https://github.com/dipy/dipy/blob/master/dipy/nn/model.py#L74) 

I know you are in the middle of both worlds (reconstruction vs neural network). However, in our current architecture, the function name  `fit` has two different goals here. 

We need to put some thinking to fix this issue.

frheault replied:
I can change predict() to __predict() et rename fit() to predict()?

The current predict function could (should) be considered private anyway, so this could work nicely.

skoudoro replied:
sounds like a good plan

skoudoro commented on dipy/nn/histo_resdnn.py:
```diff
@@ -0,0 +1,245 @@
+    logger.setLevel(level=log_level)
+
+
+class HistoResDNN(object):
```
No need of `(object)`. Can you remove it?

skoudoro commented on dipy/nn/histo_resdnn.py:
```diff
@@ -0,0 +1,245 @@
+        The model was re-trained for usage with a different basis function
+        ('tournier07') like the proposed model in [1, 2].
+
+        To obtain the pre-trained model, use:
```
`use::` instead

2 colons to avoid indentation issue in the docstring

skoudoro commented on dipy/nn/histo_resdnn.py:
```diff
@@ -0,0 +1,245 @@
+
+            # Removing negative values from the SF
+            sphere = get_sphere('repulsion724')
+            tmp_sf = sh_to_sf(sh=tmp_sh, sphere=sphere,
+                              basis_type=self.basis_type,
+                              sh_order=self.sh_order)
+            tmp_sf[tmp_sf < 0] = 0
+            tmp_sh = sf_to_sh(tmp_sf, sphere, smooth=0.0006,
+                              basis_type=self.basis_type,
+                              sh_order=self.sh_order)
```
this is strange to me. Why do you need to convert it and then come back?

frheault replied:
The model does predict negative fodf value (as SF), simply because this was not a constrained included in the original training/model.

But from SH you cannot remove negative value on SF, so you need to evaluate the SH, threshold, then fit new SH again.

Sub-optimal, yes. But necessary (I think)

skoudoro commented on doc/examples/histology_resdnn.py:
```diff
@@ -0,0 +1,158 @@
+"""
+.. histology_resdnn:
+
+==========================================
```
need to have the same size of th title

skoudoro commented on doc/examples/histology_resdnn.py:
```diff
@@ -0,0 +1,158 @@
+
+==========================================
+Local reconstruction using the Histological ResDNN
+==========================================
```
need to have the same size of the title

skoudoro commented on doc/examples/histology_resdnn.py:
```diff
@@ -0,0 +1,158 @@
+save_nifti('predicted_sh.nii.gz', predicted_sh, affine)
+
+"""
+Preparing the scene using Fury. The ODF slicer expects spherical function
```
FURY

skoudoro reviewed (commented):
Hi @frheault,

Thank for the update. See below for some comments.

FYI, you did not need to remove the doctest/docstring, you just needed to add `+skip not have_tf`.  Look at the example [here](https://github.com/dipy/dipy/blob/master/dipy/viz/projections.py#L58)

arokem commented on dipy/nn/histo_resdnn.py:
```diff
@@ -0,0 +1,245 @@
+            raise ValueError('b0 must be present for DWI normalization.')
+        logger.info('b0 indices found are: {}'.format(b0_indices))
+
+        mean_b0 = data[..., b0_indices]
```
This is confusing because it's not mean_b0 yet. Maybe combine with line below as a one-liner? 
```suggestion
        mean_b0 = np.mean(data[..., b0_indices], axis=-1)
```

arokem commented on dipy/nn/histo_resdnn.py:
```diff
@@ -0,0 +1,245 @@
+        mean_b0 = np.mean(mean_b0, axis=-1)
+
+        # Detect number of b-values and extract a single shell of DW-MRI Data
+        unique_shells = np.sort(np.unique(gtab.bvals))
```
Should use https://github.com/dipy/dipy/blob/master/dipy/core/gradients.py#L795 to be tolerant to small differences

arokem commented on dipy/nn/histo_resdnn.py:
```diff
@@ -0,0 +1,245 @@
+        flat_dw_sh_coef = dw_sh_coef[mask > 0]
+        flat_pred_sh_coef = np.zeros(flat_dw_sh_coef.shape)
+
+        count = len(flat_dw_sh_coef) // 1000
```
Do you want to make 1000 something users can control? We have something like that in DTI.

arokem reviewed (commented):
Thanks for implementing this. Can't wait to try this out. 

I had a few comments and suggestions

frheault commented on dipy/nn/histo_resdnn.py:
```diff
@@ -0,0 +1,245 @@
+            one b0)
+        mask : np.ndarray (optional)
+            Binary mask of the brain to avoid unnecessary computation and
+            unreliable prediction outside the brain.
```
It simply avoids computation for voxel that do not have any SH coefficient.
I will make it clearer in the description.

frheault commented on dipy/nn/histo_resdnn.py:
```diff
@@ -0,0 +1,245 @@
+            logger.warning('Mask should be provided to accelerate '
+                           'computation, and because predictions are '
+                           'not reliable outside of the brain.')
+            mask = np.sum(data, axis=-1)
```
I changed the description to add
```
Default: Compute prediction only for nonzero voxels (with at least
            one nonzero SH coefficient).
```
If there is noises my line will be the whole volume, if it is already skull-strip it's going to avoid computation where SH are all zeros.

frheault commented:
@skoudoro I added the `@doctest_skip_parser` at the top of the __init__ function and everything is crashing still because of the `>>>` in the doctest (that needs tensorflow). Did I do it wrong? (last commit added that)

skoudoro commented on dipy/nn/histo_resdnn.py:
```diff
@@ -0,0 +1,260 @@
+        ('tournier07') like the proposed model in [1, 2].
+
+        To obtain the pre-trained model, use::
+        >>> resdnn_model = HistoResDNN()
+        >>> fetch_model_weights_path = get_fnames('histo_resdnn_weights')
+        >>> resdnn_model.load_model_weights(fetch_model_weights_path)
```
for each line, the following comment is missing: ` # skip if not have_tf `. 

The decorator `doctest_skip_parser` will catch it and ignore it for the test.

frheault replied:
Thanks, I missed that part!

skoudoro commented:
Hi @frheault,

Thank you for the update.

Can you add Tensorflow to the coverage matrix of our CI's? You just need to add TensorFlow [here](https://github.com/dipy/dipy/blob/master/.github/workflows/coverage.yml#L16)

skoudoro commented:
Is there any additional comment for this PR?

Garyfallidis commented on doc/examples/histology_resdnn.py:
```diff
@@ -0,0 +1,157 @@
+os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'
+
+"""
+This ResDNN model requires single-shell data with b0, the data is fetched
```
with one or more b0s

Garyfallidis commented on doc/examples/histology_resdnn.py:
```diff
@@ -0,0 +1,157 @@
+
+"""
+To accelerate computation a brain mask must be computed. A median-otsu filter
+isolates the main signal and then, using a floodfill, the main 'blob' is
```
Please remove vague description about median_otsu as not necessary. Just say we use `median_otsu` to extract the brain.

Garyfallidis commented on doc/examples/histology_resdnn.py:
```diff
@@ -0,0 +1,157 @@
+picked. The resulting mask is saved for visual inspection.
+"""
+
+b0_indices = np.where(gtab.bvals == 0)[0]
```
``gtab.b0s_mask`` is the better alternative to doing this.

Garyfallidis commented on doc/examples/histology_resdnn.py:
```diff
@@ -0,0 +1,157 @@
+
+interactive = False
+sphere = get_sphere('repulsion724')
+predicted_sf = sh_to_sf(sh=predicted_sh, sphere=sphere,
```
There is a better way to do this here. As long as you provide the ``B_matrix`` in ``odf_slicer`` your can load directly the spherical harmonics (the spherical function is then saved as a memory map -- heavily reducing memory usage).

Garyfallidis reviewed (changes requested):
Thank you @frheault this is super close to be merged. Please correct the small requested changes. @skoudoro will make a quick PR on top of yours to help with memory reduction for visualization purposes.

skoudoro commented:
Here is the PR https://github.com/frheault/dipy/pull/2 @frheault, Feel free to merge it.

skoudoro commented:
Thank you @frheault for this work. All concerns have been addressed so I am going ahead a merge this PR.
